### DISCUSSION TITLE
Cluster Studio Dashboard and retain_*_columns

### DISCUSSION THREAD

**Discussion by medwar99** [moj-analytical-services/splink#1947] [Q&A]
Hi there, 

I understand that the waterfall chart, cluster studio etc. require the settings `retain_matching_columns` and `retain_intermediate_calculation_columns` to be set to True in order to pass the `_raise_error_if_necessary_waterfall_columns_not_computed()` checks and visualise the appropriate comparison information in the visualisations. 

One issue I'm finding is setting these to True is often prohibitive for large datasets and QA often only looks at a smaller representative sample, so keeping these full columns isn't needed. For example I often use the cluster graph visualisation and the node table most frequently, and so the waterfall part of the cluster studio is less often used and so those two columns aren't really needed (admittedly this is my personal usage). 

I'm wondering if it's worthwhile considering whether the comparison columns can be computed on the fly for just the records identified for the cluster studio in the `cluster_ids` list (either provided by the user or by the chosen `sampling_method`). Given that edge list could we create those missing columns, and then progress to create the result of the visualisation as needed?

I tried to call `linker.compare_two_records()` on the node pair for each identified edge with the `df_edges` of  `cluster_studio.df_edges_as_records()`, with the intent to join these back into `df_edges` in order to then pass along to the visualisation within the jinja templating. This wasn't too successful for me though, as I had a few issues getting the comparison to run due to missing term_frequency tables, although this may be something I'm missing.

Another thought is a jinja template for cluster studio which doesn't have the comparison-related requirement, only showing the graph vis and node/edge tables?

**Comment by RobinL** [comment]
Hi @medwar99

Yes - that's right, they both need to be set to true - and you're right, this results in a significant performance hit.

> I'm wondering if it's worthwhile considering whether the comparison columns can be computed on the fly for just the records identified for the cluster studio in the `cluster_ids` list (either provided by the user or by the chosen `sampling_method`)

This is a good suggestion.  I need to think a bit about how this could be implemented without adding too much complexity.

At the moment, to solve this problem 'manually', one possibility would be to:
- predict and cluster with `retain_*` options set to False
- save the node IDs of the cluster you're innterete
- filter the original data to include only those node ids
- run `predict()` and `cluster()` on that small subset of nodes with the `retain_*` options set to True

But I recognise that doesn't give you the same flexiblity (e.g. sampling by density).

One other tips that may help get you what you need without needing any changes to Splink: If you set the highest `threshold_match_probability` possible when you call `linker.predict(threshold_match_probability=x)` you will reduce the size of the dataset that needs to be stored, and therefore reduce the negative performance impact of the `retain_*` settings.

I'll raise this as an issue, but I think getting this right would be quite a bit of work, so it's probably not going to get to the top of our priorities for a while.

You're right - if you have term frequency adjustments, these need to be loaded before you can call `linker.compare_two_records()`. This isn't well dealt with in Splink right now.  There's an example of how to do it here:
https://moj-analytical-services.github.io/splink/demos/examples/duckdb/real_time_record_linkage.html

I agree that the jinja template idea might be the simplest solution in the short run to get something running
